### DISCUSSION TITLE
5.9.8 release and disabling SO_REUSEADDR changes

### DISCUSSION THREAD

**Discussion by terssec** [strongswan/strongswan#1560] [Q&A] [answered]
Firstly, apologies for my ignorance, I am not a developer as such. I suspect that changes introduced in release 5.9.8 have broken my custom script, which utilised UDP hole punching (with a rendezvous server ) to establish connections between peers behind NAT. The script would make use of SO_REUSEADDR to bind to UDP port 4500 and punch holes outbound to the rendezvous server. The script would then reuse this port binding to contact the remote peer on its public IP address (obtained from the rendezvous server). This all worked as expected in version 5.9.1. Since moving to 5.9.9 my python script returns the following exception: "[Errno 98] Address already in use" -- even when run as root. I am hoping that someone far more knowledgeable than me will be able to tell me if my suspicion is correct, and whether there's an easy way to disable this behaviour. I didn't find any compilation options that looked like they might be of help.

From the release log:

"The socket plugins don't set the SO_REUSEADDR option anymore on the IKE UDP sockets, so an error is triggered if e.g. two daemons (e.g. charon and charon-systemd) are running concurrently using the same ports."

https://github.com/strongswan/strongswan/releases/tag/5.9.8

Thank you in advance

**Comment by terssec** [comment, marked as answer]
I added the following lines of code back, recompiled, and everything is now working again on my Unix host:

https://github.com/strongswan/strongswan/commit/83da1337129221935afdbc16ffd5a6573e974073
https://github.com/strongswan/strongswan/commit/5f9ad62a8156c543868d1487e5a3d0d425c969ad

Would it be possible to add functionality to disable the new behaviour via a compile-time option, or is this just a very unusual requirement? I'm OK with making the changes and recompiling every time there's a new version, I'm just a little concerned if it could break other things (though, I don't think it should?).

**Reply by tobiasbrunner** [reply]
> Would it be possible to add functionality to disable the new behaviour via a compile-time option, or is this just a very unusual requirement?

I guess it's not very common that external programs open sockets on the same ports as the IKE daemon.  If the order in which these sockets are bound is not strictly controlled (not even sure if that 100% prevents it), you'd also run into the issue seen before when two IKE daemons opened their sockets on the same port (i.e. while both programs can send packets from that port, only the "wrong" one might receive packets on it) . I guess in your case the daemon is always started before the script.

The script you describe sounds interesting, and very similar to what strongSwan's mediation extension does (see e.g. [this test scenario](https://strongswan.org/testing/testresults/p2pnat/medsrv-psk/)).

I suppose we could add a compile- or even runtime (i.e. strongswan.conf) option.

**Reply by terssec** [reply]
Thanks Tobias. I somewhat trivialised my use case, as it's actually an entire solution I've built complete with a frontend and backend.

But yes, the "script" first starts the daemon, advertises its IP address and port, downloads other configuration data from a backend server (subnets, overlapping subnet translations, secrets, etc), and then initiates the the connection via vici. It only initiates the connection once both sides have "checked in" with the backend server, and are "ready" -- so yes there is a set order of operations. It seems to be working fine, I haven't run into any issues in the environments I've tested in.

But yes, fine, in the meantime I'll just compile without the changes to SO_REUSEADDR.

Thanks for your help and input on this.
